Support allowlisted modules and op overloads in AOTAutogradCache

jamesjwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128329

Ops in torch, torch.functional, and torch.nn.functional are cache safe by default (at least, based on my cursory audit of the ops). This fixes a few tests that use these ops with the cache.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Code reviews

[ghstack-poisoned]

jamesjwu committed:
Fixes

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

zou3519 commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -72,8 +72,13 @@ def check_node_safe(node: Node):
     Checks that the node only uses supported operators. We are starting with very
     conservative cacheability constraints, and incrementally adding more support as we expand.
     """
+    SAFE_TORCH_MODULES = ("torch.functional", "torch.nn.functional", "torch")
```
What's the definition of "safe"? That there are no global side effects?

zou3519 replied:
There's a lot of private `torch._*` APIs that I'm a bit nervous about. If you want to be conservative, then only cache on the public APIs

ezyang replied:
Are you going to also enforce some serializability constraint on the arguments to these functions?

zou3519 replied:
One category of functions to be aware of are things that save/load from files -- the contents could change from run-to-run. I would also hope that Dynamo already graph breaks on those. Another conservative thing to do here is eliminate functions with string constant arguments

bdhirsh replied:
> One category of functions to be aware of are things that save/load from files -- the contents could change from run-to-run. I would also hope that Dynamo already graph breaks on those

Do you know of any examples of this? I guess `torch.load/save`, although I would hope that dynamo graph breaks on this for the reason you said (dynamo would have to guard on the tensor metadata of what it loaded, so the output of the load would need to be a graph input).

Although in terms of being conservative or not: I guess one question is to what extent the caching logic should be free to assume dynamo guard correctness when deciding what is safe to cache or not.

>  Another conservative thing to do here is eliminate functions with string constant arguments

The cache key is computed by serializing the FX graph and converting it into a key. So any hard-coded strings and ints that are inputs to nodes will end up in the cache key too?

zou3519 replied:
> The cache key is computed by serializing the FX graph and converting it into a key. So any hard-coded strings and ints that are inputs to nodes will end up in the cache key too?

Let's assume torch.load("foo.pt") is in the FX graph (it really shouldn't be, but for the sake of demonstration). Then, this node is "unsafe" because the contents of foo.pt can change between runs.

What I meant by "eliminating functions with string constant args" is to catch cases like this -- "foo.pt" can stay the same but it's referring to some location that can change.

bdhirsh replied:
Sounds good - if we want to be overly conservative then we can definitely start by not including nodes that take in string arguments when deciding what is cache safe.

Although in the limit, we'll need to either:
* undo that and decide that we trust dynamo enough to include these in the cache
* build up an allow-list of "safe" ops that accept string inputs (e.g. `aten.div.rounding_mode`)

jamesjwu committed:
Rebase for clean tests

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Code review

[ghstack-poisoned]

jamesjwu committed:
Update on "Support allowlisted modules and op overloads in AOTAutogradCache"


Ops in torch, torch.functional, and torch.nn.functional are cache safe by default (at least, based on my cursory audit of the ops). This fixes a few tests that use these ops with the cache.


[ghstack-poisoned]

jamesjwu committed:
Update on "Support allowlisted modules and op overloads in AOTAutogradCache"


Ops in torch, torch.functional, and torch.nn.functional are cache safe by default (at least, based on my cursory audit of the ops). This fixes a few tests that use these ops with the cache.


[ghstack-poisoned]

jamesjwu committed:
One more rebase

[ghstack-poisoned]

zou3519 commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -64,8 +64,13 @@ def check_node_safe(node: Node):
     def is_torch_function(target):
+        if isinstance(target, torch._ops.OpOverload):
+            return True
+        if hasattr(target, "__module__") and target.__module__ in SAFE_TORCH_MODULES:
```
Just to leave it here in its own comment: I want a definition of "safe" in the docstring for `check_node_safe` so one can verify it. My understanding of the definition is:
- our cache key is computed from the FX graph produced by Dynamo and the input example values
- something is "safe" if the same cache key results in a compiled artifact that has the same behavior

jamesjwu replied:
Thanks, will add! Also added some tests + an extra check that the namespace for a function under a SAFE_TORCH_MODULE is not private. If it is private, we will still allow it if it's a builtin function (i.e. implemented in C++). 

There's still a danger of someone monkeypatching a public API into torch and trying to cache that, which I left on a TODO for one of the tests. But I think that edge case is too specific for us to catch here, unless there's an easy way to distinguish between them that I haven't thought of.

jamesjwu committed:
Code review: ignore private namespaced torch functions that aren't builtins

[ghstack-poisoned]

jamesjwu committed:
Add note and comment to check_node_safe docstring

[ghstack-poisoned]

zou3519 commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -63,11 +63,36 @@ def check_node_safe(node: Node):
+        # Don't blindly allow private functions in the torch namespace
+        is_private = target.__name__.startswith("_")
+        return (
+            getattr(target, "__module__", None) in SAFE_TORCH_MODULES and not is_private
```
The thing I am struggling with right now is that while it may be true that all APIs in these modules today are safe, nothing prevents someone from adding a new unsafe API and introducing silent incorrectness to the cache. For example, `torch._higher_order_ops.triton_kernel_wrapper_mutation` is an API that is not cache safe (it uses integers to refer to triton kernels), but we could easily make it a public API under torch.* some day.

bdhirsh replied:
Agreed, I don't think we would want to blindly allow anything under `torch.*` ("we might arbitrarily move higher order ops in there" feels like a good example).

How do you feel about something like this to start? Allow things under `torch.*` to be cached if either:

(1) it is a `builtin_function_or_method` (ideally, dynamo would inline all torch -related things straight to their C binding but sadly not today)

(2) it is under `torch.nn.functional` or `torch.functional` (our assumption here is that everything under these modules are things that dynamo morally should be able to trace into C-bindings today, but it does not).

That would allow us to not blanket-allow everything else under `torch.*`, like higher order ops that we happen to make public

jamesjwu replied:
Got it, I can do that; that said, I feel like it might kind of just punts the problem later, when there's a torch.* function that we actually want to support that doesn't trace into C. 

I wonder if there's a way we can check safety by being able to understand whether a function generates guards in dynamo. If there's an API we can use to check "are there no dynamo guards associated with this function", I think that would be a decent proxy for safety.

jamesjwu replied:
I'll file an issue for a more general solution to this problem but for now, I will hardcode a few torch.* functions that are safe but don't bind to C (i.e. torch.einsum) in a list for this PR. 

I think a good solution here will involve a tag in native_functions.yaml for any torch functions that are implemented in python but are cache safe

jamesjwu replied:
Update, there aren't any specific torch.* functions that actually need hard coding: torch.einsum is actually just an alias for troch.functional.einsum, so it's already covered (similar for others like it)

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu commented:
Gonna make the torch.* change, no need to review just yet

jamesjwu committed:
Remove torch.* functions from safe list

[ghstack-poisoned]

jamesjwu commented:
Update, not even gonna need an allowlist here because `torch.einsum` aliases to `torch.functional.einsum`. So we can cross the bridge of "Python implemented torch.* function that's cache safe" later if there is one we care about

jamesjwu commented:
Good for review now

bdhirsh commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -63,11 +63,36 @@ def check_node_safe(node: Node):
+        (i.e, the set of inputs that go into our cache key is sufficient to distinguish its behavior)
+
+    To accomplish this safety check, we consider the following functions to be safe:
+        - Public functions under modules torch, torch.functional, and torch.nn.functional: these are
```
I think the `"under torch"` part is out of date? (given Richard's worry about "stuff people put under `torch.foo` today that might be unsafe to cache, like HOPs).

I agree we are just kind of punting the problem - although it would probably be useful to see the cache start to miss by running it more (maybe even on the dynamo CI shard?), and see what ops actually show up in `torch.foo` that fail the cache check today?)

jamesjwu commented:
@pytorchbot rebase

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu commented:
@pytorchbot merge -i "Existing failure unrelated, fixed by  https://github.com/pytorch/pytorch/pull/129340"

jamesjwu commented:
@pytorchbot merge -i

jamesjwu commented:
@pytorchbot merge -i
